tests/perf/perf_tests.cc

```diff
@@ -203,6 +204,119 @@ static inline std::ostream& operator<<(std::ostream& os, duration d)
+
+    using print_fn = std::function<void(const column&, std::FILE *file, const result& r)>;
+
+    template <typename F>
+    column(sstring header, int prec, F fn) : header{header}, prec{prec} {
+        using result_t = std::invoke_result_t<F,const result&>;
+        constexpr auto is_integral = std::is_integral_v<result_t>;
+        constexpr auto is_double = std::is_same_v<result_t, double>;
+        constexpr auto is_duration = std::is_same_v<result_t, duration>;
+        if constexpr (is_integral || is_double || is_duration) {
+            static constexpr std::string_view fmt_str = is_double ? "{:>{}.{}f}": "{:>{}}";
+            text_printer = [=](const column& c, std::FILE *file, const result& r) {
+                fmt::print(file, fmt_str, fn(r), c.width, c.prec);
+            };
+        } else {
+            static_assert(false, "unsupported return type");
+        }
+        double_fn = [fn](const result& r) {
+            if constexpr(is_duration) {
+                return fn(r).value;
+            } else {
+                return static_cast<double>(fn(r));
+            }
+        };
+    }
+
+    void print_header(std::FILE *file, const char* str = nullptr) const {
+        fmt::print(file, "{:>{}}", str ? str : header, width);
+    }
+
+    // column header
+    sstring header;
+
+    // width for the column in text output
+    int width = 11;
+
+    // precision in case of double
+    int prec = 3;
+
+    // used by stdout and md formats to print as text
+    print_fn text_printer;
+
+    // used by json format to extract double result
+    std::function<double(const result&)> double_fn;
+};
+
+using columns = std::vector<column>;
+
+static void print_text_header(std::FILE* out,
+    const columns& cols,
+    int name_length,
+    const char* start_delim = "",   // before the line
+    const char* middle_delim = " ", // between each column
+    const char* end_delim = "",     // end of line
+    const char* test_name_header = "test",
+    const char* header_override = nullptr) {
+
+    fmt::print(out, "{}{:<{}}", start_delim, header_override ? header_override : test_name_header, name_length);
+    for (auto& c : cols) {
+        fmt::print(out, "{}", middle_delim);
+        c.print_header(out, header_override);
+    }
+    fmt::print(out, "{}\n", end_delim);
+}
+
+static void print_result_columns(std::FILE* out,
+    const columns& cols,
+    int name_length,
+    const result& r,
+    const char* start_delim = "",
+    const char* middle_delim = " ",
+    const char* end_delim = ""
+    ) {
+
+    fmt::print(out, "{}{:<{}}", start_delim, r.test_name, name_length);
+    for (auto& c : cols) {
+        fmt::print(out, "{}", middle_delim);
+        c.text_printer(c, out, r);
+    }
+    fmt::print(out, "{}\n", end_delim);
+}
+
+const std::vector<column> text_only_columns{
+    {"iterations" , 0, [](const result& r) { return r.total_iterations / r.runs; }}
+};
+
+// columns for json ouput
+const std::vector<column> json_columns{
```

Comment:
this variable is not used outside of this translation unit, so probably we can mark it with `static` ? like:
```c++
static const std::vector<column> json_columns{
  // ...
};
```